### PR TITLE
Update tests to use local ledger and setting state in template project

### PR DESCRIPTION
**Pull request by MartinMinkov**
**Description**

This PR adds better unit tests to the template zkapp project. It scaffolds using the local ledger and simulates updating the state in the local ledger. This will give users a much better starting point when they are messing around with the API.

This was also additionally done to make writing documentation for testing easier. As I was writing docs for methods for testing your smart contract, it felt weird that this sort of code wasn't already included.

**Tested By**
Updating and running the unit tests

### CODE REVIEW

File: templates/project-ts/src/index.test.ts [hunk 1]
```diff
@@ -1,20 +1,69 @@
...
+
+  afterAll(async () => {
+    // `shutdown()` internally calls `process.exit()` which will exit the running Jest process early.
+    // Specifying a timeout of 0 is a workaround to defer `shutdown()` until Jest is done running all tests.
+    // This should be fixed with https://github.com/MinaProtocol/mina/issues/10943
+    setTimeout(shutdown, 0);
+  });
+
+  it('generates and deploys the `Add` smart contract', async () => {
+    const zkAppInstance = new Add(zkAppAddress);
```
Comment (ymekuria) [inline templates/project-ts/src/index.test.ts:51]: nit: You could instantiate the zKAppInstance in the beforeEach block and avoid having to do it in every test. Up to you, either way works.

```
describe('Add', () => {
  let deployerAccount: PrivateKey,
    zkAppAddress: PublicKey,
    zkAppPrivateKey: PrivateKey,
    zkAppInstance: Add;

  beforeEach(async () => {
    await isReady;
    deployerAccount = createLocalBlockchain();
    zkAppPrivateKey = PrivateKey.random();
    zkAppAddress = zkAppPrivateKey.toPublicKey();
    zkAppInstance = new Add(zkAppAddress);
  });
```

File: templates/project-ts/src/index.test.ts [hunk 2]
```diff
@@ -1,20 +1,69 @@
...
+    await localDeploy(zkAppInstance, zkAppPrivateKey, deployerAccount);
+    const num = zkAppInstance.num.get();
+    expect(num).toEqual(Field.one);
+  });
+
+  it('correctly updates the num state on the `Add` smart contract', async () => {
+    const zkAppInstance = new Add(zkAppAddress);
```
Comment (ymekuria) [inline templates/project-ts/src/index.test.ts:58]: same as above.

File: templates/project-ts/src/index.ts [hunk 3]
```diff
@@ -18,9 +18,15 @@ import {
 export class Add extends SmartContract {
   @state(Field) num = State<Field>();
 
-  // initialization
   deploy(args: DeployArgs) {
```
Comment (ymekuria) [inline templates/project-ts/src/index.ts:21]: Do we need this deploy method still?

**Review by ymekuria — APPROVED** [review]
Great idea to add this for users! I had a couple of comments around instantiating the zkapp instance and the deploy method. Nice work!

File: templates/project-ts/src/index.test.ts (hunk 1 above)
Comment (MartinMinkov) [reply templates/project-ts/src/index.test.ts:51]: I'm running into errors when I try this, weird :/ 

```
    Check signature: Invalid signature on party 1 for key B62qrtFVa8cA4p4pk9DYT3LUn6koxFv6TCCjNbyunCQYw2NocpRnLeM
```

I think it's fine to do it this current way for now.

File: templates/project-ts/src/index.ts (hunk 3 above)
Comment (MartinMinkov) [reply templates/project-ts/src/index.ts:21]: I think it's helpful to have it here just to expose that this function exists for people first trying out SnarkyJS. In other examples, we should try to optimize it out if it's not needed

**Pull request merged**